File: LeetCodePractice/src/DP/TwoSequenceType/NumberOfLongestIncreasingSubsequence.java
package DP.TwoSequenceType;
/*Given an unsorted array of integers, find the number of longest increasing subsequence.

Example 1:
Input: [1,3,5,4,7]
Output: 2
Explanation: The two longest increasing subsequence are [1, 3, 4, 7] and [1, 3, 5, 7].
Example 2:
Input: [2,2,2,2,2]
Output: 5
Explanation: The length of longest continuous increasing subsequence is 1, and there are 5 subsequences' length is 1, so output 5.
Note: Length of the given array will be not exceed 2000 and the answer is guaranteed to be fit in 32-bit signed int.*/
public class NumberOfLongestIncreasingSubsequence {
	public int findNumberOfLIS(int[] nums) {
        int len = nums.length;
        //dp_longest[i] means LIS length ended with nums[i]
        int[] dp_longest = new int[len];
      //dp_longest[i] means # of LIS ended with nums[i]
        int[] dp_count = new int[len];
        for(int i = 0; i < len; i++){
        	dp_count[i] = 1;
        }
        for(int i = 0; i < len; i++){
        	for(int j = 0; j < i; j++){
        		if(nums[i] > nums[j]){
        			if(dp_longest[i] == dp_longest[j] + 1){
        				dp_count[i] += dp_count[j];
        			} else if(dp_longest[i] < dp_longest[j] + 1){
        				dp_count[i] = dp_count[j];
        				dp_longest[i] = dp_longest[j] + 1;
        			}
        		}
        	}
        }
        int longest = 0;
        for(int l : dp_longest){
        	longest = Math.max(longest, l);
        }
        int res = 0;
        for(int i = 0; i < len; i++){
        	if(dp_longest[i] == longest){
        		res += dp_count[i];
        	}
        }
        String s = 1 + "" + 2;
        return res;
    }
}

File: LeetCodePython/NumberOfLongestIncreasingSubsequence.py
"""
Given an integer array nums, return the number of longest increasing subsequences.

Notice that the sequence has to be strictly increasing.

Example 1:
Input: nums = [1,3,5,4,7]
Output: 2
Explanation: The two longest increasing subsequences are [1, 3, 4, 7] and [1, 3, 5, 7].

Example 2:
Input: nums = [2,2,2,2,2]
Output: 5
Explanation: The length of the longest increasing subsequence is 1, and there are 5 increasing subsequences of length 1, so output 5.


Constraints:
1 <= nums.length <= 2000
-10^6 <= nums[i] <= 10^6

analysis:
DP
TC: O(N^2)
"""
from typing import List


class NumberOfLongestIncreasingSubsequence:
    def findNumberOfLIS(self, nums: List[int]) -> int:
        res = 0
        n = len(nums)
        length = [1] * n
        count = [1] * n
        for i in range(n):
            for j in range(i):
                if nums[j] < nums[i]:
                    if length[j] + 1 > length[i]:
                        length[i] = length[j] + 1
                        count[i] = 0
                    if length[j] + 1 == length[i]:
                        count[i] += count[j]
        max_length = max(length)
        for i in range(n):
            if length[i] == max_length:
                res += count[i]
        return res